### ISSUE TITLE
feat(pkg-list): Add a `Safe` removal list

### ISSUE THREAD

**Issue by Rudxain** [Universal-Debloater-Alliance/universal-android-debloater-next-generation#583]
### Describe the feature you want

https://discord.com/channels/1168607797081022514/1168607798196715573/1273024610035564595

We should split `Recommended` in 2 variants:
- Recommended: for outright malicious apps, or with known security/privacy vulnerabilities
- Safe: for potentially useful apps

## Alternative solution
We could move some packs from `Recommended` to `Advanced`, but that seems misleading, as removing `Recommended` packs should never break the system in any way

## Prior art
There's precedent for this: [uBO](https://github.com/gorhill/uBlock) does something similar: malware/badware & ads are filter-lists that are enabled by default, but "annoyances" is disabled by default because some users want what others consider "annoying"

### Acknowledgements

- [X] This issue is not a duplicate of an existing feature request.
- [X] I have chosen an appropriate title.
- [X] All requested information has been provided properly.

**Comment by jacklollz2** [comment]
I was just about to recommend this.
Some of these "Recommended" entries are not for the average user, like removing Gmail and such.

**Comment by Rudxain** [comment]
I've been considering another alternative solution, which seems more complicated but could pay-off in the long run:

1. Move all `Advanced`s to `Expert`
2. Rename `Advanced` to `Safe`
3. Move all `Recommended`s to `Safe`
4. Gradually (opportunistically) move some `Safe` packs to `Recommend`

The 1st 3 steps can be automated, but the last one requires human work

**Comment by Frigyes06** [comment]
I agree with this. I think these packages make it into recommended because most package list maintainers (rightfully) think that they are bloat/spyware. If I understand correctly, you're proposing a list that includes only packages the user is unlikely to interact with, therefore won't likely need, and only include packages that are highly unlikely to break any other functions?

**Comment by Rudxain** [comment]
> If I understand correctly, you're proposing a list that includes only packages the user is unlikely to interact with, therefore won't likely need, and only include packages that are highly unlikely to break any other functions?

Yes! But those are just a few of many factors to take into account when deciding the recommendation category.

For example, a `Recommended` pack could be so malicious, that even if it does break the system when removed, it would be worth it. However, if the breakage is too much, then the "evilness" cancels out, so it should be in `Advanced` instead.

It could be said that my proposal/RFC only adds 1 or 2 more factors to decide the recommendation, but those factors were already being considered since a long time, so only the categories/recommendations would change

**Comment by AnonymousWP** [comment]
This sounds like an okay feature, but we need to investigate how to implement this, because it's also prone to subjectiveness. Also, our top priority for now should be 1.1.0 and the current open PRs:

- https://github.com/Universal-Debloater-Alliance/universal-android-debloater-next-generation/milestone/4
- https://github.com/Universal-Debloater-Alliance/universal-android-debloater-next-generation/pulls

**Comment by Rudxain** [comment]
Should we close #640 as dupe of this?

**Comment by crash5** [comment]
I would like to suggest a "name" key to store the app name in the uad file too. For example, the "Bolt: Request a Ride" app package name is "ee.mtakso.client". Which is pretty hard to figure out without additional googling. Some samsung packages has the same "sane" naming too :)

**Comment by Rudxain** [comment]
> "name" key

I agree, but there's multiple names:
- The display name declared in `AndroidManifest.xml` of the version installed on each device (each device can have the same pack with a different name)
- The multiple alternative names in the `description` field, which may include inferred (unofficial) names (such as "vebgm" -> "Video Editor BackGround Music")

We still haven't found a proper way to show the display-name and app-icon of installed apps

**Comment by AnonymousWP** [comment]
Will be done together with https://github.com/Universal-Debloater-Alliance/universal-android-debloater-next-generation/discussions/646 and https://github.com/Universal-Debloater-Alliance/universal-android-debloater-next-generation/discussions/608.

**Comment by Rudxain** [comment]
I think we should transfer this issue to [the new repo](https://github.com/Universal-Debloater-Alliance/universal-android-preinstalled-lists/), as it's likely we'll migrate before doing this change

**Comment by AnonymousWP** [comment]
> I think we should transfer this issue to [the new repo](https://github.com/Universal-Debloater-Alliance/universal-android-preinstalled-lists/), as it's likely we'll migrate before doing this change

But it's a pretty big change. It's part of the 2.0.0 milestone, so it'll take some time before we get there. But yeah, I definitely agree we should move this issue (along with other package-related issues) to that repo.

**Comment by Rudxain** [comment]
> packages can have more than one tag

This is also the case for [IME](https://github.com/Universal-Debloater-Alliance/universal-android-debloater-next-generation/wiki/Glossary#IME)s:
- `Recommended` if you have another IME to fill its role
- `Advanced` if there are no other IMEs

This is why [ADL has a `replace` category](https://github.com/MuntashirAkon/android-debloat-list/blob/1ac05ca509b8a5849c3256b974f9fa1e00daf856/README.md#values-for-removal). I feel like `replace` would be very helpful

**Comment by Muriilow** [comment]
This shouldn't be that large of a problem. at least for UAD. 

I already did some of the modifications, and I really didnt change much :) 

The first idea is to maintain compatibility, so even though I removed these camps in the json its possible to add in any json object. 

` 

    uad_lists.rs:20
    #[derive(Deserialize, Debug, Clone, PartialEq, Hash, Eq)]
    #[serde(rename_all = "camelCase")]
    pub struct Package {
        pub list: UadList,
        pub description: String,
        dependencies: Option<Vec<String>>, //Legacy: Not displayed on the UI
        needed_by: Option<Vec<String>>, //Legacy: Not displayed on the UI
        labels: Option<Vec<String>>, //Legacy: Not displayed on the UI
        pub removal: Removal,
    }
`

Then I also changed how the enums are organized: 

I prefer the way said by @kikecalpe in #583 

> -    Safe; for that packages that don't break nothing on any brand/device.
> -    Unsafe; for bootloops or any documented breakage on each brand/device.
> -    Untested (or Expert); for those packages that are not fully tested, but have some kind of description or info.
> -    Advanced; for packages known to be safe, but have issues on some brands/devices.

But instead of Untested or Expert I changed to Disruptive, to maintain the logic said in [FAQ](https://github.com/Universal-Debloater-Alliance/universal-android-debloater-next-generation/wiki/FAQ)

This way we have an Enum like this in uad_list:163

   pub const ALL: [Self; 6] = [
          Self::All,
          Self::Safe,
          Self::Advanced,
          Self::Disruptive,
          Self::Unsafe,
          Self::Unlisted,
   ];
In the aplication (already working on an Ubuntu Environment)
<img width="942" height="689" alt="image" src="https://github.com/user-attachments/assets/cc80d720-15af-4b31-8fc9-e633c9af0ad3" />


Im hugely against the idea of having multiple tags: 

>... a package can be "Malicious" but "Unsafe", so people can keep a close eye on those packages that cannot be removed.

I think we already have a tool to say if something is malicious and that is the description, tags should have a unique meaning "Is it safe". Havin tags that is not close to this idea makes more confusing to the user and more complex to implement. 

I also made a script in python that changes the fields "Recommended" to "Safe", and "Expert" to "Disruptive". Expert and Advanced can have similar meanings so thats not good. 

I can update the script python somewhere, but the result is like this to all the objects 

>   "com.sonymobile.apnupdater.res.overlay_305": {
>     "list": "Oem",
>     "description": "Overlay for APN Updater. Overlays are usually themes.",
>     "removal": "Disruptive"
>   },

**Comment by Muriilow** [comment]
Some notes: 
- The search is still working, even without the tag in the json. Is this updated? 
  - > labels: Will be removed in a future version, especially when [description search](https://github.com/Universal-Debloater-Alliance/universal-android-debloater-next-generation/issues/184) is implemented. 



- I want to open a new branch that follows the [Trunk-Based Development](https://github.com/Universal-Debloater-Alliance/universal-android-debloater-next-generation/blob/main/CONTRIBUTING.md) to pull this change, but before that i need my pull request to be checked so I dont have any problem with git versions and some opinions of what could change . #1113

**Comment by Rudxain** [comment]
> changes the fields "Recommended" to "Safe", and "Expert" to "Disruptive".

I agree with this. It's a small step in the right direction!

> Expert and Advanced can have similar meanings so thats not good.

I also agree that there must be more distinction (both in principle and in practice), but if we re-add `Recommended` in the future, then it'd be nice if there were less `removal` levels to worry about. That's why I proposed merging `Advanced` with `Expert`, because having 2 "middle-levels" isn't useful enough (IMO, not speaking for the team)

> The search is still working, even without the tag in the json. Is this updated?

yeah, `labels` should've been removed long ago 😅

**Comment by Muriilow** [comment]
> That's why I proposed merging Advanced with Expert, because having 2 "middle-levels" isn't useful enough (IMO, not speaking for the team) 

That's fair, I can do that. For now Im going to wait what everyone has to say and then we can continue!

**Comment by kikecalpe** [comment]
My comment was long ago, I'll try to go back my mind to explain me.

> But instead of Untested or Expert I changed to Disruptive, to maintain the logic said in [FAQ](https://github.com/Universal-Debloater-Alliance/universal-android-debloater-next-generation/wiki/FAQ)

I don't understand this logic, but this is a language handicap.
I selected that words as a suggestion because are easy and clear concepts for me, but of course there should be most descriptive ones.
I came here because some tags were confusing and not enough descriptive. About "Disruptive", yes, I had to search for its meaning 😅, but if it sounds good for you and can synthesize that concept from the packages you want to group, you know what's better for UAD. 

> I think we already have a tool to say if something is malicious and that is the description, tags should have a unique meaning "Is it safe". Havin tags that is not close to this idea makes more confusing to the user and more complex to implement.

You cannot group together by description, as a point to have in mind.
I looked for easily group packages with common points, and mentioned "Malicious" as an example and because it makes sense for me and was mentioned before, not as a request. If you look exclusively for tags to group by different levels of safety, it's good the way you mentioned.

**Comment by Muriilow** [comment]
> About "Disruptive", yes, I had to search for its meaning 😅. 

Yeah thats fair, it is an unusual word for non-native english speakers, maybe the way is just to combine both advanced and expert and set all to be advanced :p 

> If you look exclusively for tags to group by different levels of safety, it's good the way you mentioned.

 Agreed !

**Comment by iamanaws** [comment]
**On renaming Recommended to Safe:**

I agree with this change. "Recommended" sounds like advice ("you should do this"), which makes it risky because users may bulk-remove packages without reviewing them. Renaming to Safe removes that implication and makes it purely a risk assessment: "this won't break your device."

---

**On keeping Recommended as a separate category:**

I'm not convinced we should keep it. What's "malicious" or "useless" to one person may be useful to another. That's inherently subjective and doesn't belong in a tag system. The fact that a package is in the list at all already implies someone considered it removable, the tag should only answer "how risky is removal?"

---

**On `Advanced` vs `Expert`:**

I'm fine keeping both. While the boundary between them has some subjectivity, having two mid-level risk tiers provides useful granularity for the thousands of packages we manage.

| Tag | Meaning |
|-----|---------|
| `Safe` | Low risk; unlikely to break anything |
| `Advanced` | Medium risk; may affect some features |
| `Expert` | Higher risk; requires user knowledge |
| `Unsafe` | Known to cause serious issues |

I think this keeps the system simple: one axis (removal risk), four clear levels, no subjective judgments about whether an app is "good" or "bad"

**Comment by Rudxain** [comment]
I have a new idea. Back to the topic of tags/labels, they're considered "useless" because they're free-form and arbitrary (unless there's enforcement of conventions, like Wikipedia's Categories). And tags aren't as useful now that description-search is a thing, because tags are just keywords. This is the _"wild-card"/"infinite-dimensional"_ side of the spectrum.

We're currently on the _one-dimensional_ side of the spectrum, but what if we just **increment the dimension to 2** instead of allowing arbitrary dimensions? Like this:

|             |😈 Bad |🤔 Neutral |❤️ Good |
|----------|-----------|---------------|-------------|
| 🗑️ **Safe** | *REMOVE NOW!* | user pref | "If it ain't broke, don't fix it" |
| 🔬 **Expert** | weigh pros & cons | "If it ain't broke, don't fix it" | why remove? |
| ⚠️ **Unsafe** | "I'm never paying this brand again" | *DON'T remove* | _"You're a psychopath if you remove this"_ | 

These 2 dimensions solve the 2 main problems that users (and we) complain about. And best of all, they're concrete and limited!

**Comment by tristinegeer153-commits** [comment]
https://github.com/Universal-Debloater-Alliance/universal-android-debloater-next-generation/issues/1298
